Unmatched format specifier causes crashes but not compile errors

I realized a scary behavior change on compile. Code like below that has unmatched format specifiers used to cause compile errors, but we no longer see it on master because of some changes around Diag.

```diff
diff --git a/proxy/http2/Http2ClientSession.cc b/proxy/http2/Http2ClientSession.cc
index 434b4a3c4..4ada20fca 100644
--- a/proxy/http2/Http2ClientSession.cc
+++ b/proxy/http2/Http2ClientSession.cc
@@ -208,7 +208,7 @@ Http2ClientSession::new_connection(NetVConnection *new_vc, MIOBuffer *iobuf, IOB
     Debug("ssl_early_data", "read_from_early_data = %" PRId64, this->read_from_early_data);
   }
 
-  Http2SsnDebug("session born, netvc %p", this->_vc);
+  Http2SsnDebug("session born, netvc %p %s", this->_vc);
 
   this->_vc->set_tcp_congestion_control(CLIENT_SIDE);
 

```

On 9.1.x branch, the code above does not compile.
```
Http2ClientSession.cc:211:42: error: more '%' conversions than data arguments [-Werror,-Wformat-insufficient-args]
  Http2SsnDebug("session born, netvc %p %s", this->_vc);
                                        ~^
```
But on master, I don't see the compile error, and ATS crashes when I run it.
```
* thread #5, name = '[ET_NET 3]', stop reason = EXC_BAD_ACCESS (code=1, address=0x41b58ab3)
  * frame #0: 0x0000000104604df0 libclang_rt.asan_osx_dynamic.dylib`__sanitizer::internal_strlen(char const*) + 16
    frame #1: 0x00000001045ca7c5 libclang_rt.asan_osx_dynamic.dylib`printf_common(void*, char const*, __va_list_tag*) + 2069
    frame #2: 0x00000001045caa17 libclang_rt.asan_osx_dynamic.dylib`wrap_vfprintf + 103
    frame #3: 0x00000001039f0372 libtscore.10.dylib`Diags::print_va(this=0x0000612000000ac0, debug_tag="http2_cs", diags_level=DL_Debug, loc=0x000000010195ce80, format_string="[%lld] session born, netvc %p %s", ap=0x00000001092fe320) const at Diags.cc:273:7
```

The change above is something I made up to demonstrate the error, but a real error on master is already hidden by the behavior change. I see this compile error below on 9.1.x but not on master.

```
HttpSessionManager.cc:150:79: error: format specifies type 'long long' but the argument has type 'size_t' (aka 'unsigned long') [-Werror,-Wformat]
    Debug("http_ss", "Search for host name only not IP.  Pool size %" PRId64, m_fqdn_pool.count());
    ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~
../../include/tscore/Diags.h:330:39: note: expanded from macro 'Debug'
      diags->log(tag, DL_Debug, &loc, __VA_ARGS__);    \
                                      ^~~~~~~~~~~
1 error generated.
```

--- a/include/ts/apidefs.h.in
+++ b/include/ts/apidefs.h.in
@@ -51,11 +51,65 @@
 #define tsapi
 #endif
 
+/** Apply printf format string compile-time argument checking to a function.
+ *
+ *
+ * For example, given the following function from DiagsTypes.h:
+ *
+ * @code
+ * class Diags {
+ *
+ * ...
+ *
+ * void
+ * print(
+ *     const char *tag,
+ *     DiagsLevel level,
+ *     const SourceLocation *loc,
+ *     const char *fmt,
+ *     ...) const;
+ *
+ * ...
+ *
+ * };
+ * @endcode
+ *
+ * This macro can be used to apply compiler checking for ... against the fmt
+ * argument like so:
+ *
+ *
+ * @code
+ * class Diags {
+ *
+ * ...
+ *
+ * void
+ * print(
+ *     const char *tag,
+ *     DiagsLevel level,
+ *     const SourceLocation *loc,
+ *     const char *fmt,
+ *     ...) const TS_PRINTFLIKE(5, 6);
+ *
+ * ...
+ *
+ * };
+ * @endcode
+ *
+ * Note in this case, (5, 6) rather than (4, 5) is passed because `this`
+ * counts as the implicit first parameter of this member function.
+ *
+ * @param[in] fmt_index The index of the format string argument, with argument
+ * indexing starting at 1.
+ *
+ * @param[in] arg_index The index of the first format argument string, with
+ * argument indexing starting at 1.
+ */
 #if !defined(TS_PRINTFLIKE)
 #if defined(__GNUC__) || defined(__clang__)
-#define TS_PRINTFLIKE(fmt, arg) __attribute__((format(printf, fmt, arg)))
+#define TS_PRINTFLIKE(fmt_index, arg_index) __attribute__((format(printf, fmt_index, arg_index)))
 #else
-#define TS_PRINTFLIKE(fmt, arg)
+#define TS_PRINTFLIKE(fmt_index, arg_index)
 #endif
 #endif
 

--- a/include/tscore/HTTPVersion.h
+++ b/include/tscore/HTTPVersion.h
@@ -40,7 +40,7 @@ public:
 
   uint8_t get_major() const;
   uint8_t get_minor() const;
-  int get_flat_version() const;
+  uint32_t get_flat_version() const;
 
 private:
   uint8_t vmajor = 0;
@@ -82,7 +82,7 @@ HTTPVersion::get_minor() const
 /*-------------------------------------------------------------------------
   -------------------------------------------------------------------------*/
 
-inline int
+inline uint32_t
 HTTPVersion::get_flat_version() const
 {
   return vmajor << 16 | vminor;

--- a/include/tscore/LogMessage.h
+++ b/include/tscore/LogMessage.h
@@ -60,18 +60,18 @@ public:
   LogMessage(std::chrono::milliseconds throttling_interval);
 
   /* TODO: Add BufferWriter overloads for these. */
-  void diag(const char *tag, SourceLocation const &loc, const char *fmt, ...);
-  void debug(const char *tag, SourceLocation const &loc, const char *fmt, ...);
-  void status(SourceLocation const &loc, const char *fmt, ...);
-  void note(SourceLocation const &loc, const char *fmt, ...);
-  void warning(SourceLocation const &loc, const char *fmt, ...);
-  void error(SourceLocation const &loc, const char *fmt, ...);
-  void fatal(SourceLocation const &loc, const char *fmt, ...);
-  void alert(SourceLocation const &loc, const char *fmt, ...);
-  void emergency(SourceLocation const &loc, const char *fmt, ...);
+  void diag(const char *tag, SourceLocation const &loc, const char *fmt, ...) TS_PRINTFLIKE(4, 5);
+  void debug(const char *tag, SourceLocation const &loc, const char *fmt, ...) TS_PRINTFLIKE(4, 5);
+  void status(SourceLocation const &loc, const char *fmt, ...) TS_PRINTFLIKE(3, 4);
+  void note(SourceLocation const &loc, const char *fmt, ...) TS_PRINTFLIKE(3, 4);
+  void warning(SourceLocation const &loc, const char *fmt, ...) TS_PRINTFLIKE(3, 4);
+  void error(SourceLocation const &loc, const char *fmt, ...) TS_PRINTFLIKE(3, 4);
+  void fatal(SourceLocation const &loc, const char *fmt, ...) TS_PRINTFLIKE(3, 4);
+  void alert(SourceLocation const &loc, const char *fmt, ...) TS_PRINTFLIKE(3, 4);
+  void emergency(SourceLocation const &loc, const char *fmt, ...) TS_PRINTFLIKE(3, 4);
 
-  void message(DiagsLevel level, SourceLocation const &loc, const char *fmt, ...);
-  void print(const char *tag, DiagsLevel level, SourceLocation const &loc, const char *fmt, ...);
+  void message(DiagsLevel level, SourceLocation const &loc, const char *fmt, ...) TS_PRINTFLIKE(4, 5);
+  void print(const char *tag, DiagsLevel level, SourceLocation const &loc, const char *fmt, ...) TS_PRINTFLIKE(5, 6);
 
   void diag_va(const char *tag, SourceLocation const &loc, const char *fmt, va_list args);
   void debug_va(const char *tag, SourceLocation const &loc, const char *fmt, va_list args);

--- a/include/tscore/ink_apidefs.h
+++ b/include/tscore/ink_apidefs.h
@@ -54,11 +54,13 @@
 
 /*  Enable this to get printf() style warnings on the Inktomi functions. */
 /* #define PRINTFLIKE(IDX, FIRST)  __attribute__((format (printf, IDX, FIRST))) */
+/** For further information, see the doxygen comments for TS_PRINTFLIKE in
+ * include/ts/apidefs.h.in */
 #if !defined(TS_PRINTFLIKE)
 #if defined(__GNUC__) || defined(__clang__)
-#define TS_PRINTFLIKE(fmt, arg) __attribute__((format(printf, fmt, arg)))
+#define TS_PRINTFLIKE(fmt_index, arg_index) __attribute__((format(printf, fmt_index, arg_index)))
 #else
-#define TS_PRINTFLIKE(fmt, arg)
+#define TS_PRINTFLIKE(fmt_index, arg_index)
 #endif
 #endif
 

--- a/proxy/http/HttpTransact.cc
+++ b/proxy/http/HttpTransact.cc
@@ -4031,7 +4031,7 @@ HttpTransact::handle_forward_server_connection_open(State *s)
   if (real_version != s->host_db_info.app.http_data.http_version) {
     // Need to update the hostdb
     s->updated_server_version = real_version;
-    TxnDebug("http_trans", "Update hostdb history of server HTTP version 0x%x", s->updated_server_version);
+    TxnDebug("http_trans", "Update hostdb history of server HTTP version 0x%x", s->updated_server_version.get_flat_version());
   }
 
   s->state_machine->do_hostdb_update_if_necessary();

